app/Image/GdHandler.php

```diff
@@ -2,221 +2,340 @@
+	 * {@inheritDoc}
 	 */
-	public function scale(
-		string $source,
-		string $destination,
-		int $newWidth,
-		int $newHeight,
-		int &$resWidth,
-		int &$resHeight
-	): void {
-		$res = $this->readImage($source);
-		$sourceImg = $res['image'];
-		$mime = $res['mime'];
-		$width = $res['width'];
-		$height = $res['height'];
-
-		if ($newWidth === 0) {
-			$newWidth = (int) round($newHeight * ($width / $height));
-		} else {
-			$tmpHeight = (int) round($newWidth / ($width / $height));
-			if ($newHeight !== 0 && $tmpHeight > $newHeight) {
-				$newWidth = (int) round($newHeight * ($width / $height));
+	public function load(MediaFile $file): void
+	{
+		try {
+			$inMemoryBuffer = new InMemoryBuffer();
+			$this->reset();
+
+			$originalStream = $file->read();
+			if ((stream_get_meta_data($originalStream))['seekable']) {
+				$inputStream = $originalStream;
 			} else {
-				$newHeight = $tmpHeight;
+				// We make an in-memory copy of the provided stream,
+				// because we must be able to seek/rewind the stream.
+				// For example, a readable stream from a remote location (i.e.
+				// a "download" stream) is only forward readable once.
+				$inMemoryBuffer->write($originalStream);
+				$inputStream = $inMemoryBuffer->read();
 			}
-		}
-
-		$image = imagecreatetruecolor($newWidth, $newHeight);
 
-		imagecopyresampled($image, $sourceImg, 0, 0, 0, 0, $newWidth, $newHeight, $width, $height);
+			$imgBinary = stream_get_contents($inputStream);
+			rewind($inputStream);
 
-		$this->writeImage($destination, $image, $mime);
-
-		imagedestroy($image);
-		imagedestroy($sourceImg);
-
-		$resWidth = $newWidth;
-		$resHeight = $newHeight;
+			// Determine the type of image, so that we can later save the
+			// image using the same type
+			// TODO: Replace `imagecreatefromstring` by `\Safe\imagecreatefromstring` after https://github.com/thecodingmachine/safe/issues/352 has been resolved
+			error_clear_last();
+			$gdImgStat = getimagesizefromstring($imgBinary);
+			if ($gdImgStat === false) {
+				throw ImageException::createFromPhpError();
+			} else {
+				$this->gdImageType = $gdImgStat[2];
+			}
+			if (!in_array($this->gdImageType, self::SUPPORTED_IMAGE_TYPES, true)) {
+				$this->reset();
+				throw new MediaFileUnsupportedException('Type of photo is not supported');
+			}
 
-		// Optimize image
-		if (Configs::getValueAsBool('lossless_optimization')) {
-			ImageOptimizer::optimize($destination);
+			// Load image
+			error_clear_last();
+			// TODO: Replace `imagecreatefromstring` by `\Safe\imagecreatefromstring` after https://github.com/thecodingmachine/safe/issues/352 has been resolved
+			$img = imagecreatefromstring($imgBinary);
+			if ($img === false) {
+				throw ImageException::createFromPhpError();
+			}
+			$this->gdImage = $img;
+
+			// Get EXIF data to determine whether rotation is required
+			// `exif_read_data` only supports JPEGs
+			if (in_array($this->gdImageType, [IMAGETYPE_JPEG, IMAGETYPE_JPEG2000], true)) {
```

Comment:
Didn't we have a similar conversation recently (on another PR)? Do we have to rely on `exif_read_data` here rather than the preferred metadata extractor?

Reply:
Actually, it was an earlier discussion on _this_ PR: https://github.com/LycheeOrg/Lychee/pull/1351#discussion_r910573992 :smiley:

Reply:
Basically yes. We could improve on an edge case which only happens if Imagick is not available (which I deem an edge case, too). So it would be an edge case of an edge case. We could do it, but I am not sure if it is worth the effort.

When I was working on the GD handler and also fixed some bugs which apparently seem to have been there forever, I wondered whether we should drop the GD handler entirely. I noticed the bugs when I added the test cases.

The whole code of this GD handler is wicked. Even the author of the GD library discourages its use and says that there are much more potent alternatives (like Imagick). Of course I kept the GD handler, because it was not the point of the PR to drop it and I fixed the bugs. But I would not be sad to see it vanish some day even though I put a lot of work into it.

Reply:
I'm not sure how much of an edge case the lack of Imagick is. GD typically comes with PHP out-of-the-box but Imagick needs to be installed separately if I remember correctly? So personally I would prefer to keep GD support in.

The GD handler has certainly worked in the past for common cases at least (and I fixed my share of bugs in it), though recently I noticed that PHP was running out of memory when it needed to rotate a 20MP image using GD; I had to increase the limit from 128M to 256M. That doesn't happen with Imagick; either its memory management is more efficient or the core Imagick code is not subject to PHP memory limits? This is something I'll need to test, as I suspect that the new code, with its in-memory caching, will result in additional memory pressure.

BTW, have you tested if the new code is faster during imports? I would expect it to be since it doesn't reload the original image all the time. It should be especially visible with import via symlink for images requiring rotation...